fairscale/nn/moe/moe_layer.py

```diff
@@ -83,16 +75,18 @@ def forward(self, *input: Tensor, **kwargs: Any) -> Tensor:
         d_model = input[0].shape[2]
         # Reshape into S tokens by dropping sequence dimension.
         reshaped_input = input[0].reshape(-1, d_model)
-        self.l_aux, combine_weights, dispatching_mask = self.gate(reshaped_input)
-        dispatched_input = self.all_to_all_dispatch(dispatching_mask, reshaped_input)
+        self.l_aux, combine_weights, dispatch_mask = self.gate(reshaped_input)
+        dispatched_input = torch.einsum("sec,sm->ecm", dispatch_mask.float(), reshaped_input)
+        dispatched_input = _AllToAll.apply(self.group, dispatched_input)
         # Re-shape after all-to-all: ecm -> gecm
         dispatched_input = dispatched_input.reshape(self.world_size, self.num_local_experts, -1, d_model)
         chunks = dispatched_input.chunk(self.num_local_experts, dim=1)
         expert_outputs = []
         for chunk, expert in zip(chunks, self.experts):
             expert_outputs += [expert(chunk)]
         expert_output = torch.cat(expert_outputs, dim=1)
+        expert_output = _AllToAll.apply(self.group, expert_output)
```

Comment:
oh, wait - shouldn't this all_to_all come after the reshape below?

Reply:
The re-shape needs to follow the the all_to_all since it is the all_to_all that re-shaped the data. One rationale for doing this change was to put the re-shape in the right place (doesn't affect correctness).